src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,822 @@
+    bool open(const char *path) noexcept;
+
+    bool open(const char *path, const int flags) noexcept;
+
+    bool open(const char *path, const int flags, const mode_t mode) noexcept;
+
+    // throw_on_states() used for throw an exception on requested "bit"
+    // flag can be: diostates::buffer_failbit and diostates::posixio_errno
+    // Multiple bits can be passed with "|" bitwise operator also.
+    void throw_on_states(const int bit, const bool do_throw) noexcept;
+
+    // Clear all situations which marked to throw exceptions
+    void clear_throws() noexcept;
+
+    // Returns current bits/vars in stream based on diostates enum
+    int current_state() noexcept;
+
+    // Returns 1 when there is no set bits/vars otherwise 0
+    bool good() noexcept;
+
+    // Returns true or false if buffer_failbit is set or not
+    bool buffer_failure() noexcept;
+
+    // Returns 0 or POSIX errno based on posixio_errno value
+    int posixio_failure() noexcept;
+
+    // Resets failure bits/vars to false (or 0 about posixio_errno)
+    void clear() noexcept;
+
+    // Flush buffer of stream
+    // Non _nothrow variant may throw:
+    // dio::iostream_error on Buffer failure (buffer did not allocated at construction time)
+    // dio::posixio_error on POSIX I/O failure
+    ssize_t flush_nothrow() noexcept;
+    ssize_t flush();
+
+    // write() functions, Used for write many types
+    // Non _nothrow variants may throw:
+    // dio::iostream_error on Buffer failure (buffer did not allocated at construction time)
+    // dio::posixio_error on POSIX I/O failure
+    ssize_t write_nothrow(const char *msg) noexcept;
+    ssize_t write(const char *msg);
+
+    ssize_t write_nothrow(const char msg) noexcept;
+    ssize_t write(const char msg);
+
+    ssize_t write_nothrow(const std::string &msg) noexcept;
+    ssize_t write(const std::string &msg);
+
+    ssize_t write_nothrow(const int num) noexcept;
+    ssize_t write(const int num);
+
+    ssize_t write_nothrow(const unsigned num) noexcept;
+    ssize_t write(const unsigned num);
+
+    ssize_t write_nothrow(const long num) noexcept;
+    ssize_t write(const long num);
+
+    ssize_t write_nothrow(const unsigned long num) noexcept;
+    ssize_t write(const unsigned long num);
+
+    ssize_t write_nothrow(const endline &) noexcept;
+    ssize_t write(const endline &);
+
+    ssize_t write_nothrow(const flushbuf &) noexcept;
+    ssize_t write(const flushbuf &);
+
+    template <typename A, typename B, typename ...C> inline ssize_t write_nothrow(const A &a, const B &b, const C & ...c)
+            noexcept
+    {
+        if (!good()) return -1;
+        ssize_t ret1 = write_nothrow(a);
+        if (ret1 < 0) return -1;
+        ssize_t ret2 = write_nothrow(b, c...);
+        return (ret2 > 0) ? ret1 + ret2 : ret1;
+    }
+
+    template <typename A, typename B, typename ...C> inline ssize_t write(const A &a, const B &b, const C & ...c)
+    {
+        if (!good()) return -1;
```

Comment:
I don't think so, because rewriting this template with checking failures isn't ideal (`write()`checks that itself).

Reply:
I was thinking `write()` can just check the return from `write_nx()`. But actually this highlights an issue with `write_nx()`: it's not possible to tell whether everything was written (unless you know the full size in advance).

There are actually two different things you might want when you write something:

1. Regular output; to write a _complete_ message (and know if it failed).
2. To write data to a pipe where you know that partial writes and transient failures are possible, and you want to be able to try again with the remaining data. In this case it is ok if the write call doesn't write/buffer everything (but you need to know how much it did write/buffer so that you can do the rest later). This will generally be used to write from a fixed buffer (`char *`, or `string`).

I think it is worth separating these out to two different functions. Perhaps `write` for the first (similar to what there is now, but have it return error/success, instead of number of bytes buffered) and something like `write_buf` for the 2nd (and have it return the number of characters transferred to the buffer [and possibly written to the stream]).

Then, `write` should be implemented in terms of `write_nx`.

Reply:
Should we duplicate all overloads for `write_buf()` same as `write_nx()` overloads?

Reply:
No:
> This will generally be used to write from a fixed buffer (char *, or string).

That's all that's needed for `write_buf` (and that's why it's called `write_buf`: it writes a buffer).

Reply:
Done, I think It's ready for review. If it's OK just mark this as resolved.